### PR TITLE
alternative password migration for guard

### PR DESCRIPTION
**Pull request by ro0NL**
the implementation could look like

```php
public function migratePassword(
    UserInterface $user,
    $credentials,
    PasswordUpgraderInterface $passwordUpgrader,
    UserPasswordEncoderInterface $passwordEncoder
) {
    $passwordUpgrader->upgradePassword(
        $user,
        $passwordEncoder->encodePassword($user, $credentials['password'])
    );
}
```

### CODE REVIEW

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php [hunk 1]
```diff
@@ -111,9 +115,12 @@ private function authenticateViaGuard($guardAuthenticator, PreAuthenticationGuar
         }
 
         $this->userChecker->checkPreAuth($user);
-        if (true !== $guardAuthenticator->checkCredentials($token->getCredentials(), $user, $this->userProvider instanceof PasswordUpgraderInterface ? $this->userProvider : null)) {
+        if (true !== $guardAuthenticator->checkCredentials($token->getCredentials(), $user)) {
             throw new BadCredentialsException(sprintf('Authentication failed because %s::checkCredentials() did not return true.', \get_class($guardAuthenticator)));
         }
+        if ($this->userProvider instanceof PasswordUpgraderInterface && $guardAuthenticator instanceof PasswordMigratingInterface) {
+            $guardAuthenticator->migratePassword($user, $token->getCredentials(), $this->userProvider, $this->passwordEncoder);
```
Comment (ro0NL) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php:122]: we can actually call `$this->passwordEncoder->needsRehash()` first now

**Comment by ro0NL** [comment]
also `function needsRehash(UserInterface $user, string $encoded)` is a weird signature, we should use `$user->getPassword()` instead of `$encoded` AFAIK (as does `isPasswordValid()`)

File: src/Symfony/Bundle/SecurityBundle/Resources/config/guard.xml [hunk 2]
```diff
@@ -29,6 +29,7 @@
             <argument /> <!-- User Provider -->
             <argument /> <!-- Provider-shared Key -->
             <argument /> <!-- User Checker -->
+            <argument type="service" id="security.password_encoder" />
```
Comment (ro0NL) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/guard.xml:32]: i tend to believe `DaoAuthenticationProvider` could also inject the `UserPasswordEncoder`, to save a few lines (currently it uses the factory directly)

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php (hunk 1 above)
Comment (ro0NL) [reply src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php:122]: done.

File: src/Symfony/Bundle/SecurityBundle/Resources/config/guard.xml (hunk 2 above)
Comment (ro0NL) [reply src/Symfony/Bundle/SecurityBundle/Resources/config/guard.xml:32]: we already inject the factory since 3.4 here, not worth it

**Comment by nicolas-grekas** [comment]
Thank you for the inspiration, I'm taking a slightly different option but this has been definitely useful!

**Pull request closed without merging**